### DISCUSSION TITLE
Double locale in URL still works

### DISCUSSION THREAD

**Discussion by FlorisdeG** [vercel/next.js#36856] [Help] [answered]
Hello everyone, we have a multi-locale website, but we notice that if we write a double locale in the URL the page still works. Here is an example [link](https://www.iodigital.com/nl/en). The locale for this page is `en`, even though the slug starts with a `/nl`.

What we want is for this to either redirect to `/nl` or it should give a 404 with the `nl` locale. How can we accomplish this?

**Comment by icyJoseph** [comment, marked as answer]
Mmm I think it has to do with the `[[...slug]]` catch-all route you have. 

For example, https://www.iodigital.com/nl/en/nl resolves as `nl`. The `asPath` is interesting, as it becomes `/en/nl`

I think you might need to match the resolved URL, or the `asPath`, or however is it that you process slug, and check that there's only one locale...

Not saying this should work, but something like:

```ts
const allLocales = new Set(['en', 'nl']); // not sure how, but you can get access to these in various server side methods
// slug is list in [[...slug]]
const localesSent = slug.filter(segment => allLocales.has(segment))

if (localesSent.length >= 1) { return {notFound:true}}; 

```

A lot of pseudocode there, but I guess you could try that. It is meant to reject slugs with more than one locale included.

`_middleware` would be another option.

### Edit

I hacked this together, hopefully it can guide you:

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";
import { useRouter } from "next/router";

const Home: NextPage = () => {
  const { locale } = useRouter();

  return <div>Home {locale}</div>;
};

export const getServerSideProps = async (
  context: GetServerSidePropsContext
) => {
  const allLocales = new Set(context.locales);

  const slug = context.query?.slug ?? [];

  if (!Array.isArray(slug)) return { notFound: 404 };

  const totalLocales = slug.filter((seg) => allLocales.has(seg));

  if (totalLocales.length >= 1) return { notFound: 404 };

  return { props: {} };
};

export default Home;
```

With a pages folder structure:

```shell
pages/
├── [[...slug]].tsx
├── _app.tsx
└── api
    └── hello.ts
```

Then `localhost:3000`, `localhost:3000/en` and `localhost:3000/nl` work fine, using this config:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  i18n: {
    locales: ["en", "nl"],
    defaultLocale: "en",
  },
};

module.exports = nextConfig;
```

---

As a side note, catch all routes, at the root, are savage! They really do catch all, even apparently malformed paths. Gotta be careful with that.

**Reply by FlorisdeG** [reply]
If I tried `/nl/en/nl` it would take give a 404 with `en` as the locale, not `nl`.

I fixed it now, thank you. Do you know why it accepts 2 languages? I don't really see a use case for that.

**Reply by icyJoseph** [reply]
I don't think it accepts two locales, it is more about the nature of a root level `[[...slug]]`, which are very aggressive, that when combined with `locale` URLs, place the responsibility on the developer to check the incoming `slug` value.

On a past project, we had such root level catch all, and before reaching 1.0, almost weekly, something strange happened with it.
